receiver/sqlserverreceiver/scraper.go

```diff
@@ -299,3 +326,243 @@ func (s *sqlServerScraperHelper) recordDatabaseStatusMetrics(ctx context.Context
 
 	return errors.Join(errs...)
 }
+
+func (s *sqlServerScraperHelper) recordDatabaseQueryTextAndPlan(ctx context.Context, topQueryCount uint) (plog.Logs, error) {
+	// Constants are the column names of the database status
+	const dbPrefix = "db."
+	const totalElapsedTime = "total_elapsed_time"
+	const rowsReturned = "total_rows"
+	const totalWorkerTime = "total_worker_time"
+	const queryHash = "query_hash"
+	const queryPlanHash = "query_plan_hash"
+	const logicalReads = "total_logical_reads"
+	const logicalWrites = "total_logical_writes"
+	const physicalReads = "total_physical_reads"
+	const executionCount = "execution_count"
+	const totalGrant = "total_grant_kb"
+	const queryText = "query_text"
+	const queryPlan = "query_plan"
+	rows, err := s.client.QueryRows(ctx)
+	if err != nil {
+		if !errors.Is(err, sqlquery.ErrNullValueWarning) {
+			return plog.Logs{}, fmt.Errorf("sqlServerScraperHelper failed getting rows: %w", err)
+		}
+		s.logger.Warn("problems encountered getting log rows", zap.Error(err))
+	}
+	var errs []error
+
+	totalElapsedTimeDiffs := make([]int64, len(rows))
+
+	for i, row := range rows {
+		queryHashVal := hex.EncodeToString([]byte(row[queryHash]))
+		queryPlanHashVal := hex.EncodeToString([]byte(row[queryPlanHash]))
+
+		elapsedTime, err := strconv.ParseInt(row[totalElapsedTime], 10, 64)
+		if err != nil {
+			s.logger.Info(fmt.Sprintf("sqlServerScraperHelper failed getting rows: %s", err))
+			errs = append(errs, err)
+		} else {
+			// we're trying to get the queries that used the most time.
+			// caching the total elapsed time (in millisecond) and compare in the next scrape.
+			if cached, diff := s.cacheAndDiff(queryHashVal, queryPlanHashVal, totalElapsedTime, elapsedTime/1000); cached && diff > 0 {
+				totalElapsedTimeDiffs[i] = diff
+			}
+		}
+	}
+
+	// sort the rows based on the totalElapsedTimeDiffs in descending order,
+	// only report first T(T=topQueryCount) rows.
+	rows = sortRows(rows, totalElapsedTimeDiffs, topQueryCount)
+
+	// sort the totalElapsedTimeDiffs in descending order as well
+	sort.Slice(totalElapsedTimeDiffs, func(i, j int) bool { return totalElapsedTimeDiffs[i] > totalElapsedTimeDiffs[j] })
```

Comment:
Would it be better performance wise to use a tree based/heap structure here instead of an array to have it sorted twice?

As this method is part of the hot path, it would be good to see if this could be improved, or if it is needed.

Reply:
changed to priority queue